### PR TITLE
20210211 zrq volume mounts

### PR DESCRIPTION
**Pull request by Zarquan**
- Upgraded worker nodes from small to medium.
- Added a mount of the extra local disc. 
- Moved Spark and Hadoop temp directories onto local disc. 
- Tweaked the Spark and Yarn settings to match the new configuration (needs more work).
- Recorded some performance numbers for comparison with existing configuration.

### CODE REVIEW

File: notes/zrq/20210212-01-speed-tests.txt [hunk 1]
```diff
@@ -0,0 +1,881 @@
...
+    ....
+
+    good/bad select
+    ....
+
+    RandomForestClassifier - 10% data 500 trees
+    Took 12 min 4 sec. Last updated by gaiauser at February 12 2021, 5:11:23 AM.
+
+
+Live deployment #2
```
Comment (stvoutsin) [inline notes/zrq/20210212-01-speed-tests.txt:163]: Was this run soon after the first one? Wondering any possible caching might have an effect on the result

File: notes/zrq/20210212-01-speed-tests.txt [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+    >   ***** hive-site.xml *****
+    >   hive.tez.container.size=14336
+    >   hive.tez.java.opts=-Xmx11264m
+    >   hive.auto.convert.join.noconditionaltask.size=3758096000
+
+
+# -----------------------------------------------------
+# -----------------------------------------------------
+
+dev deployment #1
```
Comment (stvoutsin) [inline notes/zrq/20210212-01-speed-tests.txt:323]: Can we add some info describing the dev settings here? i.e. Spark/Hadoop settings & info on VMs

File: notes/zrq/20210212-01-speed-tests.txt [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+        default settings, 10% data, 500 trees
+
+        main select statement
+        Took 28 min 32 sec. Last updated by gaiauser at February 12 2021, 4:03:51 AM.
+
+        RandomForestClassifier - 10% data 500 trees
+        Took 14 min 58 sec. Last updated by gaiauser at February 12 2021, 4:19:07 AM.
+
+    #
+    # Tweaked the Hadoop/Yarn settings ..
```
Comment (stvoutsin) [inline notes/zrq/20210212-01-speed-tests.txt:336]: Some quick info on what was tweaked?

File: notes/zrq/20210212-01-speed-tests.txt [hunk 4]
```diff
@@ -0,0 +1,881 @@
...
+
+        Nulls
+        Took 15 min 54 sec. Last updated by gaiauser at February 13 2021, 8:18:59 AM.
+
+    TODO
+
+        retry with caching enabled
+        retry with 1000 and 5000 trees
+
+
```
Comment (stvoutsin) [inline notes/zrq/20210212-01-speed-tests.txt:881]: Am I correct in thinking we are getting inconsistent & inconclusive results? Looks like the speed of the tests can vary quite a bit, and it looks like they are being affected by something outside of our control?

File: experiments/hadoop-yarn/ansible/22-config-spark-master.yml [hunk 5]
```diff
@@ -49,20 +50,23 @@
...
+            spark.executor.cores           4
+            spark.executor.instances      11
+
+            spark.eventLog.enabled      true
+            spark.driver.maxResultSize 8192m
+
+            spark.local.dir            {{sptemplink}}
+            spark.eventLog.dir         hdfs://{{hdhost}}:9000/spark-log
+
+
```
Comment (stvoutsin) [inline experiments/hadoop-yarn/ansible/22-config-spark-master.yml:69]: Do we have a benchmark that shows these optimizations are infact improving performance that we can point to?

Comment (stvoutsin) [reply experiments/hadoop-yarn/ansible/22-config-spark-master.yml:69]: Wondering if the performance optimization changes might have been better off on a separate branch than the mounts. (Although I realize that they are related and mounts do affect performance)

**Review by stvoutsin — APPROVED** [review]

File: notes/zrq/20210212-01-speed-tests.txt (hunk 1 above)
Comment (Zarquan) [reply notes/zrq/20210212-01-speed-tests.txt:163]: Yes - caching is an issue. Timings for different runs varies quite a lot.
Nigel's notebook has a call to clear the cache, but it doesn't clear everything. 
 
    %spark.pyspark
    # clear any previously cached data in the context ....
    sqlContext.clearCache()

File: notes/zrq/20210212-01-speed-tests.txt (hunk 2 above)
Comment (Zarquan) [reply notes/zrq/20210212-01-speed-tests.txt:323]: All of the settings are in the Ansible scripts.
- `live deployment` is the existing live deployment as it was at the time
- `dev deployment` is a new deployment using the updated Ansible scripts

File: notes/zrq/20210212-01-speed-tests.txt (hunk 3 above)
Comment (Zarquan) [reply notes/zrq/20210212-01-speed-tests.txt:336]: My bad - didn't make a note of them.

File: notes/zrq/20210212-01-speed-tests.txt (hunk 4 above)
Comment (Zarquan) [reply notes/zrq/20210212-01-speed-tests.txt:881]: Yes, absolutley.
Turns out the Ceph file store was having to cope with multiple disc failures at the time.
On a good day this notebook is limited by the available read data rate from the Ceph file store, this was a bad day and the number were all over the place.
Following these tests I contacted John Garbutt and he ended up removing a couple of discs and re-building the RAID array.

File: experiments/hadoop-yarn/ansible/22-config-spark-master.yml (hunk 5 above)
Comment (Zarquan) [reply experiments/hadoop-yarn/ansible/22-config-spark-master.yml:69]: Your tests from `20210211-ML-Notebook-Benchmarking` showed that using local disc was a lot faster, but the tiny and small VMs didn't have enough local disc to cache the whole data set.
The aim of these changes was to update the deployment to use medium VMs, which have a different disc layout, different number of cores and different amount of memory.
The configuration changes were the minimum needed to get it to perform to the same level as the results you reported for local disc in `20210211-ML-Notebook-Benchmarking` but with enough memory and disc space to handle the full data set. 
There is indeed a lot of scope for performance optimization in the future.

**Pull request merged**